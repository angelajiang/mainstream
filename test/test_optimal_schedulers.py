import os
import shutil
import subprocess
import pytest

@pytest.mark.unit
@pytest.mark.parametrize("scheduler_type", ["hifi", "stems"])
def test_optimality(scheduler_type):
    output_dir = os.path.join("test", "tmp", scheduler_type)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    schedules_dir = os.path.join(output_dir, "schedules")
    if not os.path.exists(schedules_dir):
        os.makedirs(schedules_dir)
    other_file = os.path.join(schedules_dir, scheduler_type + ".sim.debug.v0")
    exhaustive_file = os.path.join(schedules_dir, "exhaustive.sim.debug.v0")

    subprocess.check_call("test/exhaustive_vs_x_scheduler.sh " + scheduler_type, shell=True)

    other_f1s = []
    with open(other_file, "r") as f:
        for line in f:
            vals = line.split(',')
            metric = round(float(vals[1]), 2)
            other_f1s.append(metric)

    exhaustive_f1s = []
    with open(exhaustive_file, "r") as f:
        for line in f:
            vals = line.split(',')
            metric = round(float(vals[1]), 2)
            exhaustive_f1s.append(metric)

    assert other_f1s == exhaustive_f1s

    shutil.rmtree(output_dir)


@pytest.mark.unit
@pytest.mark.parametrize("scheduler_type,num_apps_range", [
    ("hifi", 5),
    ("stems", 5),
    ("exhaustive", 3),
    ("stems_cpp", 5),
])
def test_regression(regtest, scheduler_type, num_apps_range):
    output_dir = os.path.join("test", "tmp", scheduler_type)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    schedules_dir = os.path.join(output_dir, "schedules")
    if not os.path.exists(schedules_dir):
        os.makedirs(schedules_dir)
    other_file = os.path.join(schedules_dir, scheduler_type + ".sim.debug.v0")

    subprocess.check_call("test/run_scheduler_with_setups.sh {} {}".format(scheduler_type, num_apps_range), shell=True)

    with open(other_file) as f:
        for line in f:
            line = line.split(",")[:-1]
            regtest.write(",".join(line) + "\n")

    shutil.rmtree(output_dir)
