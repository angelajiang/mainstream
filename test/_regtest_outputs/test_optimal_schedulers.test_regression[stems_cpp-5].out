2,0.133914,45,45,4,3,100
2,0.126213,45,45,4,3,100
2,0.0841593,33,33,4,3,100
2,0.260686,45,45,4,3,100
2,0.196902,33,33,4,2,100
3,0.0711344,33,33,33,3,4,2,100
3,0.21522,33,33,33,4,3,3,100
3,0.185122,33,45,45,4,3,4,100
3,0.0326526,39,33,39,3,4,4,100
3,0.0481635,39,33,39,4,4,3,100
4,0.168909,69,51,57,69,4,4,4,4,100
4,0.118502,51,57,69,69,4,4,4,4,100
4,0.120092,63,63,51,51,3,4,4,3,100
4,0.0733753,45,57,51,57,4,4,2,3,100
4,0.292296,45,51,51,51,4,2,3,2,100
5,0.200025,75,75,69,51,51,4,4,4,4,2,100
4,0.154495,45,39,69,69,4,2,4,4,100
5,0.285933,57,63,63,63,63,4,4,2,3,4,100
5,0.183529,57,75,63,57,75,4,3,3,4,4,100
5,0.229862,57,75,75,63,51,4,4,4,4,2,100
